### PR TITLE
Security - Using an ORM/ODM library and validating input to prevent query injection and other attacks

### PR DESCRIPTION
**Pull request by BrunoScheufler**
This pull request is still in a stage of work-in-progress and will be completed in the next days. If you have any additions or suggestions, don't hesitate to comment!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -689,11 +689,13 @@ Developers in the project may not follow consistent code security practices, lea
...
 
-**TL;DR:** To prevent SQL/noSQL injection attacks *always* make use of an ORM/ODM or a database library that escapes or supports named or indexed parameter binding, and takes care of validating user input for expected types. **Never** use JS template strings or string concatenation to inject values into queries.
+**TL;DR:** To prevent SQL injection and other malicious attacks, _always_ make use of an ORM/ODM or a database library that escapes data or supports named or indexed parameterized queries, and takes care of validating user input for expected types. **Never** just use JavaScript template strings or string concatenation to inject values into queries as this opens your application to a wide spectrum of vulnerabilities.
 
-**Otherwise:** Un-validated user input could lead to operator injection when working with MongoDB for noSQL, and unescaped use of proper ORM/ODM will allow easy SQL injection attacks.
+**Otherwise:** Unvalidated or unsanitized user input could lead to operator injection when working with MongoDB for NoSQL, and not using a proper ORM/ODM will easily allow SQL injection attacks.
```
Comment (i0natan) [inline README.md]: @BrunoScheufler The first part was about ODMs and Mongo. Nice. Shouldn't the second paragraph emphasize sql injections? maybe bring an example: Otherwise an attacker could put an infinite JS loop in a parameter and ... (is this possible? hmm no idea)

No need to explain the context "not using a ..." because it's obvious that the otherwise relates to the TLDR

https://www.owasp.org/index.php/Testing_for_NoSQL_injection

File: sections/security/ormodmusage.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+### One Paragraph Explainer
+
+When creating your database logic you should watch out for eventual injection vectors that could be exploited by potential attackers. Writing database queries manually or not including data validation for user requests are the easiest methods to allow for these vulnerabilities. This situation is however easy to avoid when you use suitable packages for validating input and handling database operations. In many cases your system will be safe and sound by using a validation library like
+[joi](https://github.com/hapijs/joi) or [yup](https://github.com/jquense/yup) and an ORM/ODM from the list below. This should guarantee the use of parameterized queries and data bindings to ensure the validated data is properly escaped and handled without opening unwanted attack vectors. Many of these libraries will ease your life as a developer by enabling many useful features like not having to write complex queries manually, supplying types for language-based type systems or converting data types to wanted formats. To conclude, __always__ validate any data you are going to store and use proper data-mapping libraries to handle the dangerous work for you.
```
Comment (i0natan) [inline sections/security/ormodmusage.md]: @BrunoScheufler You sure this is true? Joi was meant to assert schema correctness, you can assert the some field must include string but I don't think it can identify vulenerable string patterns

File: sections/security/ormodmusage.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+### Libraries
+
+- [TypeORM](https://github.com/typeorm/typeorm)
+- [sequelize](https://github.com/sequelize/sequelize)
+- [mongoose](https://github.com/Automattic/mongoose)
+- [waterline](https://github.com/balderdashy/waterline)
```
Comment (i0natan) [inline sections/security/ormodmusage.md]: @BrunoScheufler Knex and Objection.js are more popular than Waterline and typeorm, kindly include those

File: sections/security/ormodmusage.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+🔗 [OWASP SQL Injection](https://www.owasp.org/index.php/SQL_Injection)
+
+🔗 [OWASP SQL Injection Prevention Cheat Sheet](https://www.owasp.org/index.php/SQL_Injection_Prevention_Cheat_Sheet)
+
+🔗 [Testing for NoSQL Injection](https://www.owasp.org/index.php/Testing_for_NoSQL_injection)
+
+### What other bloggers say
+
+To be added
```
Comment (i0natan) [inline sections/security/ormodmusage.md]: I liked this:
"NoSQL injection attacks may execute in different areas of an application than traditional SQL injection. Where SQL injection would execute within the database engine, NoSQL variants may execute during within the application layer or the database layer, depending on the NoSQL API used and data model. Typically NoSQL injection attacks will execute where the attack string is parsed, evaluated, or concatenated into a NoSQL API call."
From https://www.owasp.org/index.php/Testing_for_NoSQL_injection

File: sections/security/ormodmusage.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+### Libraries
+
+- [TypeORM](https://github.com/typeorm/typeorm)
+- [sequelize](https://github.com/sequelize/sequelize)
+- [mongoose](https://github.com/Automattic/mongoose)
+- [waterline](https://github.com/balderdashy/waterline)
+
+### Example
+
+To be added
```
Comment (i0natan) [inline sections/security/ormodmusage.md]: @BrunoScheufler An example of noSql attack is not very common, could be nice

File: sections/security/ormodmusage.md (hunk 2 above)
Comment (BrunoScheufler) [reply sections/security/ormodmusage.md]: You can validate types and other related things to that, although I agree it's not meant for vulnerable string patterns but rather as a first step of validating a request or user-submitted data. Sanitization libraries might be the next step in this workflow though, I could add that

File: sections/security/ormodmusage.md (hunk 3 above)
Comment (BrunoScheufler) [reply sections/security/ormodmusage.md]: @i0natan of course! will do!

File: sections/security/ormodmusage.md (hunk 4 above)
Comment (BrunoScheufler) [reply sections/security/ormodmusage.md]: @i0natan great! Thanks for helping out with that 😄

File: sections/security/ormodmusage.md (hunk 5 above)
Comment (BrunoScheufler) [reply sections/security/ormodmusage.md]: @i0natan will look into finding something for that 👍
